lib/manager/npm/extract/yarn.ts

```diff
@@ -7,30 +7,77 @@ export type YarnLock = Record<string, string> & {
   '@renovate_yarn_integrity'?: boolean;
 };
 
+interface ParseRangeReturnType {
+  protocol: string | null;
+  source: string | null;
+  selector: string | null;
+  params: string | null;
+}
+
+// https:// github.com/yarnpkg/berry/blob/b128988379a1f257b2406ebf5058b67861403bd4/packages/yarnpkg-core/sources/structUtils.ts#L260-L303
```

Comment:
I understand why you would copy a couple of functions, instead of depending on `@yarnpkg/core`, but I'm not sure it complies with the terms of the license. `@yarnpkg/core` is [licensed under the BSD 2-clause license](https://github.com/yarnpkg/berry/blob/master/LICENSE.md), which includes the following bullet-point

>Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

Reply:
oh I did not know we could use @yarnpkg/core directly ! I will make the changes :)